spec/services/sign_in/token_params_validator_spec.rb

```diff
@@ -0,0 +1,144 @@
+      }.compact
+    end
+
+    let(:grant_type) { nil }
+    let(:code) { nil }
+    let(:code_verifier) { nil }
+    let(:client_assertion) { nil }
+    let(:client_assertion_type) { nil }
+    let(:assertion) { nil }
+
+    context 'when grant_type is AUTH_CODE_GRANT' do
+      let(:grant_type) { SignIn::Constants::Auth::AUTH_CODE_GRANT }
+      let(:code) { 'some-code' }
+      let(:code_verifier) { 'some-code-verifier' }
+
+      let(:expected_response) do
+        {
+          grant_type:,
+          code:,
+          code_verifier:,
+          client_assertion:,
+          client_assertion_type:
+        }.compact
+      end
+
+      context 'when client_assertion_type is nil' do
+        let(:client_assertion_type) { nil }
+
+        it_behaves_like 'valid params'
+
+        context 'when code is missing' do
+          let(:code) { nil }
+
+          it_behaves_like 'invalid params', "Code can't be blank"
+        end
+
+        context 'when code_verifier is missing' do
+          let(:code_verifier) { nil }
+
+          it_behaves_like 'invalid params', "Code verifier can't be blank"
+        end
+      end
+
+      context 'when client_assertion_type is present' do
+        let(:client_assertion_type) { 'some-client-assertion-type' }
+
+        context 'when client_assertion_type is CLIENT_ASSERTION_TYPE' do
+          let(:client_assertion_type) { SignIn::Constants::Auth::CLIENT_ASSERTION_TYPE }
+
+          context 'when client_assertion is present' do
+            let(:client_assertion) { 'some-client-assertion' }
+
+            it_behaves_like 'valid params'
+
+            context 'when code is missing' do
+              let(:code) { nil }
+
+              it_behaves_like 'invalid params', "Code can't be blank"
+            end
+
+            context 'when code_verifier is missing' do
+              let(:code_verifier) { nil }
+
+              it_behaves_like 'valid params'
+            end
+          end
+
+          context 'when client_assertion is missing' do
+            let(:client_assertion) { nil }
+
+            it_behaves_like 'invalid params', "Client assertion can't be blank"
+          end
+        end
+
+        context 'when client_assertion_type is not CLIENT_ASSERTION_TYPE' do
+          let(:client_assertion_type) { 'invalid-client-assertion-type' }
+
+          it_behaves_like 'invalid params', 'Client assertion type is not valid'
+        end
+      end
+    end
+
+    context 'when grant_type is JWT_BEARER_GRANT' do
+      let(:grant_type) { SignIn::Constants::Auth::JWT_BEARER_GRANT }
+
+      let(:expected_response) do
+        {
+          grant_type:,
+          assertion:
+        }.compact
+      end
+
+      context 'when assertion is present' do
+        let(:assertion) { 'some-assertion' }
+
+        it_behaves_like 'valid params'
+
+        context 'when assertion is missing' do
+          let(:assertion) { nil }
+
+          it_behaves_like 'invalid params', "Assertion can't be blank"
```

Comment:
this really doesn't matter either way, but I wonder if it would be slightly more clear to have a let defined:
`let(:expected_error message) { "Assertion can't be blank" }`
and then add that to the behaves like?
`it_behaves_like 'invalid params', expected_error_message`